scripts/rollup/packaging.js

```diff
@@ -165,15 +167,80 @@ function copyNodePackageTemplate(packageName) {
     // We already created this package (e.g. due to another entry point).
     return Promise.resolve();
   }
-  // TODO: verify that all copied files are either in the "files"
-  // whitelist or implicitly published by npm.
-  return asyncCopyTo(npmFrom, to).then(() =>
-    Promise.all([
-      asyncCopyTo(resolve(`${from}/package.json`), `${to}/package.json`),
-      asyncCopyTo(resolve(`${from}/README.md`), `${to}/README.md`),
-      asyncCopyTo(resolve('./LICENSE'), `${to}/LICENSE`),
-    ])
-  );
+  // Get all entry points in the package root
+  // Exceptions: *.fb.js
+  const packageEntries = glob.sync('!(*.fb).js', {
+    cwd: from,
+  });
+  const npmFiles = fs.readdirSync(npmFrom);
+  packageEntries.forEach(entry => {
+    if (!npmFiles.includes(entry)) {
+      // Terminate the build if any entry point(Exception: *.fb.js)
+      // does not have an equivalent in ./npm.
+      console.error(
+        `Entry point ${entry} in package ${
+          packageName
+        } does not have an equivalent in ./npm`
+      );
+      process.exit(1);
+    }
+  });
+  const packageJson = `${from}/package.json`;
+  const whitelist = fs.existsSync(packageJson) && require(packageJson).files;
+  let promisesForForwardingModules = [];
+  if (!whitelist) {
+    // Terminate the build if 'files' field is missing from package.json.
+    console.error(
+      `'files' field is missing from package.json in package ${packageName}`
+    );
+    process.exit(1);
+  }
+  fs.mkdirSync(to);
+  let existingDirCache = {};
+  // looping through entries(single file / directory / pattern) in `files`
+  const whitelistedFiles = whitelist.reduce((list, pattern) => {
```

Comment:
This logic is a bit complicated. Is there any module we can use for this? Or maybe there's some simpler way to write it? At least let's extract it to a separate function.

Reply:
Thanks @gaearon, some modules may help with bits and pieces, but I'm a little hesitant including any as new dev dependencies, considering we already use things like `ncp` and `glob`(which the other modules are build upon). Also `glob` composes a whiteListedFiles list used for checking all entry points are whitelisted. If a module doing things like `copy with glob`, we'll have to run `glob` again to get the list.
[node-mkdirp](https://github.com/substack/node-mkdirp) can help creating nested directories. Do you think I can add it as dev dependency?
I will extract it to a separate function, make the code flow more intuitive.

Reply:
node-mkdirp seems fine. Why didn't we need it before?